server/src/shared/query/pagination.ts

```diff
@@ -0,0 +1,50 @@
+import { IDeferredJoinParams, IKeysetParams } from 'shared/types/pagination';
+import db from 'shared/database';
+
+export async function deferredJoinPagination({
```

Comment:
Function `deferredJoinPagination`:

Parameters: params object:
- `tableName`: name of the table containing data to be paginated
- `indexedColumnName`: name of the column within the table that is going to be criterion for ordering; it must be indexed in order for the query to make sense
- `selectAttributes`: list of attribute names that are going to be selected; if not specified, all attributes are selected
- `limit`: number of entries per page
- `orderByDirection`: direction of ordering, `'asc'` for ascending or `'desc'` for descending
- `pageNumber`: page number :)

Defined variables:
- `offset` - number of rows to be discarded in order to fetch specific page, calculated as `(pageNumber - 1) * limit`
- `innerQueryName` - a name of the inner query for fetching a page of entries containing only indexed column values
- `attributes` - table attributes to be selected; if `selectAttributes` is specified, it is equal to `selectAttributes`, otherwise, it is equal to `'*'`

The query:

```ts
return db
    .with(innerQueryName, db =>
      db
        .select(indexedColumnName)
        .from(tableName)
        .orderBy(indexedColumnName, orderByDirection)
        .limit(limit)
        .offset(offset)
    )
```

This represents the inner query in form of CTE. The inner query fetches the specified page containing only indexed column values. This way, it is not necessary to do table access for each row when applying select, then discard; that operation is applied only to values stored as index, which makes the query more performant.

```ts
.select(attributes)
    .from(tableName)
    .join(
      innerQueryName,
      `${innerQueryName}.${indexedColumnName}`,
      '=',
      `${tableName}.${indexedColumnName}`
    )
    .orderBy(`${tableName}.${indexedColumnName}`, orderByDirection);
```

The second part does table access and fetches the remaining attributes only for those entries selected by the inner query, which is accomplished with performing join operation between inner join result and the specified table.